core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -211,7 +212,8 @@ public void sendMessage(ReqT message) {
     // For unary requests, we don't flush since we know that halfClose should be coming soon. This
     // allows us to piggy-back the END_STREAM=true on the last message frame without opening the
     // possibility of broken applications forgetting to call halfClose without noticing.
-    if (!unaryRequest) {
+    if (!unaryRequest && (onReadyThreadId == -1
+        || onReadyThreadId != Thread.currentThread().getId())) {
```

Comment:
Just a thought: Consider separating this out into a method:

```java
if (needToFlushAfterSending())
```
You could move the docs to the method and expand them to discuss `onReadyThreadId`.